link/kprobe.go

```diff
@@ -156,21 +227,20 @@ func pmuKprobe(symbol string, ret bool) (*perfEvent, error) {
 	// Ensure the string pointer is not collected before PerfEventOpen returns.
 	runtime.KeepAlive(sp)
 
-	// Kernel has perf_kprobe PMU available, initialize perf event.
+	// Kernel has perf_[k,u]probe PMU available, initialize perf event.
 	return &perfEvent{
 		fd:    internal.NewFD(uint32(fd)),
 		pmuID: et,
 		name:  symbol,
-		typ:   kprobeEventType(ret),
+		typ:   typ.PerfEventType(ret),
 	}, nil
 }
 
-// tracefsKprobe creates a trace event by writing an entry to <tracefs>/kprobe_events.
+// tracefsProbe creates a trace event by writing an entry to <tracefs>/[k,u]probe_events.
 // A new trace event group name is generated on every call to support creating
-// multiple trace events for the same kernel symbol. A perf event is then opened
+// multiple trace events for the same kernel or userspace symbol. A perf event is then opened
 // on the newly-created trace event and returned to the caller.
-func tracefsKprobe(symbol string, ret bool) (*perfEvent, error) {
-
+func tracefsProbe(typ probeType, symbol, path string, offset uint64, ret bool) (*perfEvent, error) {
```

Comment:
Providing `tracefsKprobe()` and `tracefsUprobe()` helpers that wrap `tracefsProbe()` with the correct incantation of `offset` and `path` would remove a ton of churn from the tests. It would also allow us to better document what happens under the covers. (this docstring currently doesn't mention `offset` or `path`, for example)

Reply:
👍🏻 I did this for `tracefsProbe` and `pmuProbe` for consistency; I don't think it makes much sense to do the same for `createTraceFSProbeEvent` since in non-test code is called from a generic function. The only pro I see is that it would make tests (maybe) more readable

Reply:
but let me know if I should change it too

Reply:
True, the gains might not outweigh adding the helpers in this case.